### PR TITLE
Add support to create FQL X Functions

### PR DESCRIPTION
**Pull request by pnwpedro**
### Changes
* Add new command, `fqlx`. e.g. sls fqlx deploy
* Add `fqlx` command schema.
* Add support for FQL X Functions.
* Add `created_by_serverless_plugin` string value for FQL X command. This allows the plugins to take over management so that `sls fauna` removal scenarios won't interfere with `sls fqlx` and the reverse.

### Behavior
* Resource creates and updates are handled in single transaction prior to any resource removals
* Resource removals (as part of a `sls fqlx deploy` command) are handled after creates and updates. Deletes may require pagination, so they may be effected over 1 or more transactions.
* In the event that a create/update query is too large to evaluate by Fauna and returns an error, users will be required to break their infrastructure up into separate serverless files. This allows users to have a clear understanding of which resources will be grouped into transactions.

### Testing
* Added a number of integration tests for FQL X
* Verified that the previous command works as expected.
* Validated with a local fauna instance and some serverless.yml definitions

example def
```
service: boring-service

plugins:
  - "@fauna-labs/serverless-fauna"

provider: # dummy values
  name: aws
  runtime: nodejs14.x

fqlx:
  deletion_policy: retain

  client:
    secret: ${env:FAUNA_SECRET, "secret"}
    endpoint: ${env:FAUNA_ENDPOINT, "http://localhost:8443"}

  functions:
    AlwaysOne:
      body: "_ => 1"

    AlwaysThree:
      body: "_ => 3"

    AlwaysTree:
      body: "_ => 'tree'"
```

create:
```
Fauna: FQL X schema create/update transaction in progress...
Fauna: function: AlwaysOne created
Fauna: function: AlwaysThree created
Fauna: function: AlwaysTree created
Fauna: FQL X schema remove transactions in progress...
✨  Done in 4.14s.
```

noop:
```
Fauna: FQL X schema create/update transaction in progress...
Fauna: FQL X schema remove transactions in progress...
✨  Done in 4.14s.
```

after an edit:
```
Fauna: FQL X schema create/update transaction in progress...
Fauna: function: AlwaysTree updated
Fauna: FQL X schema remove transactions in progress...
✨  Done in 2.98s.
```

removal:
```
Fauna: FQL X schema create/update transaction in progress...
Fauna: FQL X schema remove transactions in progress...
Fauna: function: AlwaysTree deleted
✨  Done in 2.79s.
```

and nice error formatting:
```
Fauna: FQL X schema create/update transaction in progress...
Fauna: invalid_query: The query failed 1 validation check
---
error: Expected end-of-input
at *query*:17:8
   |
17 | return results
   |        ^
   |
✨  Done in 3.63s.
```

### CODE REVIEW

File: fqlx/queries/function.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Creates or updates a function according to the passed parameters.
+ *
+ * @param params The params to pass to the create or update call.
+ * @returns An FQL X Query
+ */
+const createUpdateFunction = (params) => {
+  return fql`
+  let results = if (Function.byName(${params.name}) != null) {
```
Comment (macmv) [inline fqlx/queries/function.js:12]: The `!= null` will break somewhat soon, once we land support for named null documents. I don't think there's any action here, just something to lookout for.

File: fqlx/queries/utils.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  return Object.entries(obj)
+    .map(([k, v]) => fql`${fqlIdent(variable)}.${fqlIdent(k)} == ${v}`)
+    .reduce((result, curr) => {
+      if (result == null) {
+        return curr
+      } else {
+        return fql`${result} && ${curr}`
+      }
+    }, null)
+}
```
Comment (macmv) [inline fqlx/queries/utils.js:43]: This seems like it could produce long queries. I would go with something like this in FQLX:
```js
Object.entries(${obj}).every(v => v[1] == f[v[0]])
```
Object.entries will give you something like this:
```js
[
  ["name", "Scout"],
  ["species", "Dog"],
]
```
and then `every` will iterate over the array, and will return `true` if _every_ predicate evaluation returned `true`.

File: fqlx/queries/queryBuilder.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+}) => {
+  const queries = [
+    ...functions.map(f => createUpdateFunction(f)),
+  ]
+
+  const reduced = queries
+    .filter((q) => !!q)
+    .reduce((res, curr) => {
+      return fql`${res}\n\n${curr}`
+    }, fql`let results = []`)
```
Comment (macmv) [inline fqlx/queries/queryBuilder.js:15]: I think you can avoid appending each time, if each `createUpdateFunction` returns one result:
```js
const queries = [
  ...functions.map(f => createUpdateFunction(f))
];

return fql`${queries}`
```

then `createUpdateFunction` would just return an object like this:
```js
  fql`
  if (Function.byName(${params.name}) != null) {
    let f = Function.byName(${params.name})
    if (${and("f", params)}) {
      { type: "function", name: ${params.name}, result: "not changed" }
    } else {
      f.update(${params})
      { type: "function", name: ${params.name}, result: "updated" }
    }
  } else {
    Function.create(${params})
    { type: "function", name: ${params.name}, result: "created" }
  }`
```

if any of the queries need to return multiple results, then feel free to leave it as-is.

**Review by macmv — COMMENTED** [review]
If you can use the `fqlx` and `fauna` commands at the same time, then I think this approach works. For example, I should be able to `sls deploy` this file:
```
fqlx:
  deletion_policy: retain

  client:
    secret: ${env:FAUNA_SECRET, "secret"}
    endpoint: ${env:FAUNA_ENDPOINT, "http://localhost:8443"}

  functions:
    AlwaysOneFQLX:
      name: "AlwaysOne"
      body: "_ => 1"

fauna:
  deletion_policy: retain

  client:
    secret: ${env:FAUNA_SECRET, "secret"}
    endpoint: ${env:FAUNA_ENDPOINT, "http://localhost:8443"}

  functions:
    AlwaysOneFQL4:
      name: "AlwaysOne"
      body: "AlwaysOneFQL4.fql"
```

and it should deploy two functions.

It would be nice to disallow duplicate function/collection/etc names between fql4/fqlx, but I don't think that's possible. The `fauna_api_version` field seems like it might be unnecessary? Everything in fql4 and fqlx shares the same namespace, so if the two plugins conflict, they'll already be editing the same document. At that point, I think it becomes a race for which plugin creates the document first, which seems messy.

Wrapping everything up into one transaction is great! It's good to start the new version with the right architecture in place.

Implementation wise, I have a couple of comments about the queries produced, and I think they could be a bit simpler. Aside from that, it looks good to me.

File: fqlx/queries/utils.js (hunk 2 above)
Comment (pnwpedro) [reply fqlx/queries/utils.js:43]: Sweet. This is great! I missed that we support Object.entries() natively.

File: fqlx/queries/queryBuilder.js (hunk 3 above)
Comment (pnwpedro) [reply fqlx/queries/queryBuilder.js:15]: Each should ideally do one thing, so this seems like a nice improvement

**Comment by pnwpedro** [comment]
@macmv, thanks for the thoughtful review! For context, I'm trying to be minimally invasive with this addition because we lack adequate test coverage.

> If you can use the fqlx and fauna commands at the same time

You can call these commands separately, `sls fauna deploy` and `sls fqlx deploy`, but it's fine if schemas are defined in the same file. However, `sls deploy` will not deploy both today because I didn't hook into the main deploy command for `fqlx`. I tried at first, but couldn't test it because `sls deploy` doesn't work for me–it tries to create an AWS stack because that's the dummy provider I'm including.

> I should be able to sls deploy this file

Again, I had trouble testing the general `sls deploy` because fauna isn't a valid provider and when using another provider it tries to do strange things. I'm wondering if we should rethink supporting that without being a valid provider.

> It would be nice to disallow duplicate function/collection/etc names between fql4/fqlx, but I don't think that's possible.

I think we could if we refactored to use a single deploy command that handles both schemas, or if each deploy command handled both schemas. However this is pretty invasive, and we'd have to increase test coverage for both APIs to make a change like this.

> The fauna_api_version field seems like it might be unnecessary? Everything in fql4 and fqlx shares the same namespace, so if the two plugins conflict, they'll already be editing the same document. At that point, I think it becomes a race for which plugin creates the document first, which seems messy.

I'm not sure whether you're talking yourself out of it being unnecessary, but in order to do away with the version, each deploy command would have to handle both schemas to determine if it should destroy a resource. Again, I think this would be require a lot more effort in the v4 side. `fauna_api_version` feels like a straight-forward way to guarantee that resources that were last created/updated by a command are managed by that command from a _removal_ point of view. Either can take over that resource by declaring it under the desired the schema.

---

In summary, I think your suggestions are really good, but I'm hesitant to support them because of LOE. With this implementation, we've isolated most changes to a completely new API, which I like a lot.

**Comment by pnwpedro** [comment]
> I had trouble testing the general `sls deploy` at first, but think I sorted that out. It looks like the general deploy command does not deploy the fqlx definition even though I tried to hook into it.

The `deploy:deploy` hook gets overridden by the last one we init, so I'll remove it from the fqlx command for now to avoid confusion

File: fqlx/client.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  const cacheKey = `${secret}/${endpoint}`
+
+  if (!clients[cacheKey]) {
+    clients[cacheKey] = new Client({
+      secret,
+      endpoint: new URL(endpoint),
+    })
+  }
+
+  return clients[cacheKey]
```
Comment (ptpaterson) [inline fqlx/client.js:15]: What is this cache for? The underlying http2 client for the fqlx is a singleton class, so multiple Clients will still use the same http2 instance. The http2 client keeps a map of open sessions keyed by endpoint. All that is to say, if you call `new Client()` with a given endpoint that has already been used, it will already reuse the existing session for that endpoint.

Comment (pnwpedro) [reply fqlx/client.js:15]: This cache is copy/paste from the v4 client.js in this library. I can remove it

File: fqlx/queries/remove.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  let names = resources.map(f => f.name)
+  return fql`
+  {
+    let mod = ${module}
+    mod.where(
+      f =>
+        !${names}.includes(f.name) &&
+        f.data?.deletion_policy != "retain" &&
+        f.data?.created_by_serverless_plugin == "fauna:v10"
+    ).order(.name).map(f => {
```
Comment (macmv) [inline fqlx/queries/remove.js:22]: This `order` call is going to collect up the whole set. So, you could add `toArray()` to the end of this, and it would have no performance impact, and it would make the client code much simpler (you could avoid paginating entirely).

On the other hand, we might fix `order` to do the right thing, but that won't be any time soon. Since you already have the pagination setup, I would leave it as-is, but if pagination ever becomes difficult to work with on the client, this can easily be switched to an array with no downsides.

File: __tests__/commands/FQLXCommands.test.js [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+    }
+  }
+
+  beforeAll(async () => {
+    const p = clientConfig.port
+    const ep = `${clientConfig.scheme}://${clientConfig.domain}${p ? ":" + p : ""}`
+    faunaClient = new Client({
+      secret: clientConfig.secret,
+      endpoint: new URL(ep),
+    })
```
Comment (macmv) [inline __tests__/commands/FQLXCommands.test.js:65]: why make a new client every time? it seems like this would just make tests take longer.

**Review by macmv — APPROVED** [review]
I have a few minor comments, but aside from those LGTM.

The tests timing out is definitely weird, so imma look into that, but this looks code-complete aside from that.

File: commands/FQLXCommands.js [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+      const q = deployQuery(this.adapt({functions}))
+      // Example expected data:
+      // res.data -> [ { type: "function", name: "MyFunc", result: "created" } ]
+      const res = await this.client.query(q)
+
+      res.data.forEach(record => {
+        if (record.result !== "noop") {
+          this.logger.success(`${record.type}: ${record.name} ${record.result}`)
+        }
+      })
```
[the remark this answers is not in the document]
Comment (pnwpedro) [reply commands/FQLXCommands.js:69]: Yeah, I was thinking about that too. I can improve this in coming iterations

File: fqlx/queries/remove.js (hunk 5 above)
Comment (pnwpedro) [reply fqlx/queries/remove.js:22]: Ah, good to know, ty ty
